prometheus/src/main/java/com/uber/m3/tally/prometheus/PrometheusReporter.java

```diff
@@ -0,0 +1,484 @@
+            Summary.Builder builder = Summary.build()
+                    .name(name)
+                    .help(String.format("%s summary", name))
+                    .ageBuckets(ageBuckets)
+                    .maxAgeSeconds(maxAgeSeconds)
+                    .labelNames(collectionToStringArray(ttags.keySet()));
+            defaultQuantiles.forEach(builder::quantile);
+            Summary summary = builder.register(registry);
+            registeredSummaries.put(collectorName, summary);
+        }
+        registeredSummaries.get(collectorName)
+                .labels(collectionToStringArray(ttags.values()))
+                .observe(interval.getSeconds());
+    }
+
+    private void reportTimerHistogram(String name, Map<String, String> tags, Duration interval) {
+        Map<String, String> ttags = tags;
+        if (tags == null) {
+            ttags = Collections.emptyMap();
+        }
+        String collectorName = canonicalMetricId(name, ttags.keySet());
+        if (!registeredHistograms.containsKey(collectorName)) {
+            Histogram histogram = Histogram.build()
+                    .name(name)
+                    .help(String.format("%s histogram", name))
+                    .buckets(defaultBuckets)
+                    .labelNames(collectionToStringArray(ttags.keySet()))
+                    .register(registry);
+            registeredHistograms.put(collectorName, histogram);
+        }
+        registeredHistograms.get(collectorName)
+                .labels(collectionToStringArray(ttags.values()))
+                .observe(interval.getSeconds());
+    }
+
+    /**
+     * Generates a canonical MetricID for a given name+label keys, not values.
+     *
+     * @param name    metric name.
+     * @param tagKeys label keys.
+     * @return canonical metric ID in a form foo+tag1=1,tag2=1.
+     */
+    static String canonicalMetricId(String name, Set<String> tagKeys) {
+        if (name == null) {
+            throw new IllegalArgumentException("metric name cannot be null");
+        }
+        Map<String, String> tags;
+        if (tagKeys == null) {
+            tags = Collections.emptyMap();
+        } else {
+            tags = new HashMap<>(tagKeys.size());
+            for (String key : tagKeys) {
+                tags.put(key, METRIC_ID_KEY_VALUE);
+            }
+        }
+        return keyForPrefixedStringMaps(name, Collections.singletonList(tags));
+    }
+
+    /**
+     * Generates a unique key for a prefix and a list of maps containing tags.
+     * <p>
+     * If a key occurs in multiple maps, keys on the right take precedence.
+     *
+     * @param prefix prefix for unique key.
+     * @param tags   list of maps containing tags.
+     * @return unique key in a format prefix+key1=value1,key2=value2
+     * @throws IllegalArgumentException when {@code prefix} is null.
+     */
+    private static String keyForPrefixedStringMaps(String prefix, List<Map<String, String>> tags) {
```

Comment:
Why does this take `List<Map<String, ...` instead of just `Map<String, String>`? At a glance it looks like the only invocation uses a single element list, making those roughly equivalent?

Reply:
Also, seems like this function may already be implemented in this library? IIUC the M3 reporter would also need it--maybe we should factor that out somewhere.

Reply:
yeah, you're right. I wanted to make go-lang and java tally libraries to be as close as possible and in [go-lang implementation it uses an array of maps. ](https://github.com/uber-go/tally/blob/master/key_gen.go#L64) 
I'll just use `Map<String, String> instead`.

Reply:
Also I couldn't find any similar functions in the codebase